In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


In [3]:
df = pd.read_csv("/content/mobile_phone_dataset.csv")

In [4]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

from statsmodels.stats.outliers_influence import variance_inflation_factor

In [5]:
features = [
    'RAM_GB',
    'Storage_GB',
    'Battery_mAh',
    'Rear_Camera_MP',
    'Front_Camera_MP'
]

target = 'Price_INR'

data = df[features+[target]].dropna()
X = df[features]
y = df[target]

In [6]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)



In [7]:
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

In [8]:
y_pred = model.predict(X_test)
print(y_pred)

[ 98850.66027301  71763.4568632   50597.76468504  42019.41753118
  71124.30061885  87506.54664188  68343.03257345  62546.02934169
  56230.26126395 108014.82896771  73485.10450423  45410.41731596
  86254.75652542  86124.93702337  95430.23598325 109816.05719807
  65142.53986092  62408.70521954  63919.69881378 118027.47384465
 111981.28336291  53621.83396254 103337.30172188 102047.84168805
  52187.53623947  86254.75652542 111405.82875249  98588.82948843
  77404.25951118 122979.41609043  52887.2448581  133886.21005192
  61182.49355569 104842.81598343  81611.49118852 121148.7633551
  54752.15950236 100545.53155014  82186.94579894 118890.65576028
  58311.94542375  76329.3143162  133023.02813628  70854.2396419
  81323.76388331  74977.26188098  57467.23380295  71412.02792406
  78109.95886552  99014.23028096  76546.08383174  66465.10847194
  68295.76120726  56278.65466473  73453.68407137  75526.76065154
  90913.52343046  68374.96440933 110045.53090488  55230.0758118
  46416.48884757 110933.7985

In [9]:
r2 = r2_score(y_test,y_pred)
print("r2 score:",r2)

r2 score: 0.9843029031245688


QUESTION 2

In [10]:
coeff_df = pd.DataFrame({
    'Feature': X.columns,
    'Coefficient': model.coef_
})
print(coeff_df)
print("Intercept:",model.intercept_)

           Feature  Coefficient
0           RAM_GB  2522.228877
1       Storage_GB   121.242688
2      Battery_mAh     3.179543
3   Rear_Camera_MP    77.736916
4  Front_Camera_MP    35.965913
Intercept: 5767.656094480262


 **RAM_GB (2522.23): Every additional 1 GB of RAM increases the predicted phone price by approximately ₹2522.
 **Storage_GB (121.24): Every additional 1 GB of storage increases the predicted price by about ₹121.
 **Battery_mAh (3.18): Each extra 1 mAh of battery capacity increases the predicted price by around ₹3.18.
 **Rear_Camera_MP (77.74): Every additional 1 MP in the rear camera increases the predicted price by approximately ₹77.74.
 **Front_Camera_MP (35.97): Every additional 1 MP in the front camera increases the predicted price by about ₹35.97.
 **Intercept (5767.66):This is the predicted base price when all feature values are zero.


 QUESTION 3

In [11]:
from statsmodels.stats.outliers_influence import variance_inflation_factor
import pandas as pd

vif_data = pd.DataFrame()
vif_data["Feature"] = X.columns
vif_data["VIF"] = [variance_inflation_factor(X.values, i) for i in range(X.shape[1])]

print(vif_data)

           Feature        VIF
0           RAM_GB   5.207077
1       Storage_GB   2.875042
2      Battery_mAh  10.706551
3   Rear_Camera_MP   3.580398
4  Front_Camera_MP   3.713949


The VIF values show that **Battery_mAh** has a VIF of **10.71**, which indicates high multicollinearity. **RAM_GB** has a VIF of **5.21**, suggesting moderate multicollinearity. The remaining features have VIF values below 5 and do not show significant multicollinearity.


*** QUESTION 4 ***

Based on the regression coefficients, **RAM_GB** is the most influential feature in predicting mobile phone price because it has the largest coefficient value (2522.23). This means that increasing RAM by 1 GB results in the largest increase in the predicted price compared to the other features.


QUESTION 5

In [12]:
X_new = X.drop(columns=['Battery_mAh'])

X_train_new, X_test_new, y_train_new, y_test_new = train_test_split(
    X_new, y, test_size=0.2, random_state=42
)

model_new = LinearRegression()
model_new.fit(X_train_new, y_train_new)

y_pred_new = model_new.predict(X_test_new)

r2_new = r2_score(y_test_new, y_pred_new)

print("New R² Score:", r2_new)

coef_df_new = pd.DataFrame({
    'Feature': X_new.columns,
    'Coefficient': model_new.coef_
})

print(coef_df_new)
print("Intercept:", model_new.intercept_)

New R² Score: 0.9735529499914821
           Feature  Coefficient
0           RAM_GB  2492.170729
1       Storage_GB   121.303798
2   Rear_Camera_MP    76.035755
3  Front_Camera_MP    38.155359
Intercept: 21973.100895552896


After removing **Battery_mAh**, the regression model was re-trained using the remaining features. The new model showed similar performance while reducing multicollinearity in the dataset. Removing the highly collinear feature made the model more stable and improved the interpretability of the regression coefficients.
